url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/EvalTrace.java

```diff
@@ -0,0 +1,921 @@
+        }
+
+        public String getName()
+        {
+            return name;
+        }
+
+        public void setEnabled(boolean enabled)
+        {
+            this.enabled = enabled;
+        }
+
+        public boolean breakHere(Eval2.EvalStack aStack)
+        {
+            return enabled && aStack.size() >= depth;
+        }
+
+        @Override
+        public String toString()
+        {
+            StringBuilder lBuilder = new StringBuilder();
+            lBuilder.append(this.name).append(", ").append("stack depth break: '").append(depth).append("'").append(enabled?", enabled":", paused");
+            return lBuilder.toString();
+        }
+    }
+
+    public static class BreakpointNot
+    implements IBreakpoint
+    {
+        private String name;
+        private boolean enabled = true;
+        private IBreakpoint bp;
+
+        public BreakpointNot(String aName, IBreakpoint aBp)
+        {
+            this.name = aName;
+            this.bp = aBp;
+        }
+
+        public String getName()
+        {
+            return name;
+        }
+
+        public void setEnabled(boolean enabled)
+        {
+            this.enabled = enabled;
+        }
+
+        public boolean breakHere(Eval2.EvalStack aStack)
+        {
+            return enabled && !(bp.breakHere(aStack));
+        }
+
+        @Override
+        public String toString()
+        {
+            StringBuilder lBuilder = new StringBuilder();
+            lBuilder.append(this.name).append(", ").append("not-composite").append(enabled?", enabled":", paused");
+            lBuilder.append("\n\t").append(bp.toString().replace("\n\t", "\n\t\t"));
+            return lBuilder.toString();
+        }
+    }
+
+    public static class BreakpointAnd
+    implements IBreakpoint
+    {
+        private String name;
+        private boolean enabled = true;
+        private List<IBreakpoint> bps;
+
+        public BreakpointAnd(String aName, List<IBreakpoint> aBps)
+        {
+            this.name = aName;
+            this.bps = aBps;
+        }
+
+        public String getName()
+        {
+            return name;
+        }
+
+        public void setEnabled(boolean enabled)
+        {
+            this.enabled = enabled;
+        }
+
+        public boolean breakHere(Eval2.EvalStack aStack)
+        {
+            boolean lBreak = enabled;
+            for(IBreakpoint lBp: bps)
+            {
+                lBreak = lBreak && lBp.breakHere(aStack);
+                if(!lBreak) return lBreak;
+            }
+            return lBreak;
+        }
+
+        @Override
+        public String toString()
+        {
+            StringBuilder lBuilder = new StringBuilder();
+            lBuilder.append(this.name).append(", ").append("and-composite").append(enabled?", enabled":", paused");
+            for(IBreakpoint lBp: bps)
+                lBuilder.append("\n\t * ").append(lBp.toString().replace("\n\t", "\n\t\t"));
+            return lBuilder.toString();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 342:346, 388:392.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.